### ISSUE TITLE
DPL-333 As a dev (Seena N) I would like to decide on how to share STAN styling with Traction so that we keep the styling up to date.

### ISSUE THREAD

**Issue by TWJW-SANGER** [sanger/traction-ui#743]
**Description**
As a dev (Seena N) I would like to decide on how to share STAN styling with Traction so that we keep the styling up to date.

There a number of options for sharing the styling between projects; we want to research these and find the option that best fits the following criteria:

- Allows each project keep up to date with changes made to the styling for other projects.
- Allows each project to explicitly use a specific version of the styling to avoid unexpected breakages. i.e. projects would explicitly have to decide to pull in the latest changes so they can be tested as part of normal processes.
- Minimises duplication of code between projects.
- Has a low complexity & maintenance overhead.

We would like to time-box this activity to 1 day of effort.

**Who the primary contacts are for this work**
Steven I, Tom W.

**Knowledge or Stake holders**
Steven I, Seena N, Sangeetha B

**Additional context or information**
We only anticipate sharing styling information for this story not logic (not full React or Vue components).

Possible options are:
- Having some kind of shared repo for styling
- Compiling Tailwind styles and committing the output to individual projects
- Something else...

Other teams on the internet seem to be exploring the same kind of issues:
https://github.com/tailwindlabs/discuss/issues/228
http://billpatrianakos.me/blog/2015/03/16/share-css-between-related-projects-with-git-subtrees/

Some guidance from Tailwind here: https://tailwindcss.com/docs/reusing-styles

**Comment by stevieing** [comment]
[Sharing STAN styling files with Traction.docx](https://github.com/sanger/traction-ui/files/8520852/Sharing.STAN.styling.files.with.Traction.docx)

**Comment by stevieing** [comment]
https://docs.google.com/document/d/1T6z1Azwu2m9mfTXEPhLfjjvusczUu8edgYrukOXNbw4/edit?usp=sharing

**Comment by harrietc52** [comment]
Just reading up on Git Subtree's which are new to me! But I think they sound like my preference at the moment. 

A few thoughts taken from [this](https://gist.github.com/SKempin/b7857a6ff6bddb05717cc17a44091202) SubTree explanation:

"It isn't readily apparent that part of the main repo is built from a subtree" - could all the files have a prefix to make it clean

"This `add` command adds all the code and files into the main repository locally; it's not just a reference to a remote repo." - just wondering where would these files get added too / would we need to manually move the files to the correct location they are required for each respective parent repo.  I guess the subtree code could be in modules that get included, but they might still need to be moved somewhere?

"You can't easily list the subtrees in your project" - might we only ever need just the 1 subtree?

"the logs are slightly confusing" [here](https://gist.github.com/SKempin/b7857a6ff6bddb05717cc17a44091202?permalink_comment_id=3959851#gistcomment-3959851) - they do look particularly confusing! but not sure much we can do about that!

**Comment by seenanair** [comment]
> Just reading up on Git Subtree's which are new to me! But I think they sound like my preference at the moment.
> 
> A few thoughts taken from [this](https://gist.github.com/SKempin/b7857a6ff6bddb05717cc17a44091202) SubTree explanation:
> 
> "It isn't readily apparent that part of the main repo is built from a subtree" - could all the files have a prefix to make it clean
> 
> "This `add` command adds all the code and files into the main repository locally; it's not just a reference to a remote repo." - just wondering where would these files get added too / would we need to manually move the files to the correct location they are required for each respective parent repo. I guess the subtree code could be in modules that get included, but they might still need to be moved somewhere?
> 
> "You can't easily list the subtrees in your project" - might we only ever need just the 1 subtree?
> 
> "the logs are slightly confusing" [here](https://gist.github.com/SKempin/b7857a6ff6bddb05717cc17a44091202?permalink_comment_id=3959851#gistcomment-3959851) - they do look particularly confusing! but not sure much we can do about that!
@harrietc52 Thank you very much for your feedback and the thoughts shared. I  have similar thoughts as well about Subtree as the first choice. The

**Comment by harrietc52** [comment]
Ah I didn't appreciate that is all that is needed! 
Yes, definitely worth weighing up whether two files justifies this added complexity. Maybe the npm module approacg is more lightweight and appealing. In the google doc [here](https://docs.google.com/document/d/1T6z1Azwu2m9mfTXEPhLfjjvusczUu8edgYrukOXNbw4/edit?usp=sharing) for npm module cons it says "But customisation required in our case and that makes the process complex", what does this mean?

**Comment by seenanair** [comment]
> @seenanair Ah I didn't appreciate that is all that is needed! Yes, definitely worth weighing up whether two files justifies this added complexity. Maybe the npm module approach is more lightweight and appealing. In the google doc [here](https://docs.google.com/document/d/1T6z1Azwu2m9mfTXEPhLfjjvusczUu8edgYrukOXNbw4/edit?usp=sharing) for npm module cons it says "But customisation required in our case and that makes the process complex", what does this mean/ do you know how come it is more complex?

@harrietc52 We cannot use the tailwind config file as such from the STAN because of framework differences (React used in STAN and Vue in  Traction). So some minor changes required for configuration file like
- Replacing purge section with .vue extension
- Removing STAN specific plugins from plugins section

**Comment by JamesGlover** [comment]
We could potentially keep the shared tailwind config in a separate file, and then import it into `tailwind.config.js` where we apply the local modifications.

Something like:

```javascript
import defaultOptions from 'style-subtree/tailwind.config.js'

module.exports = {
  ...defaultOptions,
  purge: {
    ...defaultOptions.purge,
    content: ["./src/**/*.{html,vue}"],
  },
  important: true,
  corePlugins: {},
  plugins: [],
};
```

**Comment by seenanair** [comment]
> We could potentially keep the shared tailwind config in a separate file, and then import it into `tailwind.config.js` where we apply the local modifications.
> 
> Something like:
> 
> ```js
> import defaultOptions from 'style-subtree/tailwind.config.js'
> 
> module.exports = {
>   ...defaultOptions,
>   purge: {
>     ...defaultOptions.purge,
>     content: ["./src/**/*.{html,vue}"],
>   },
>   important: true,
>   corePlugins: {},
>   plugins: [],
> };
> ```
> 
> You could probably take a similar approach with the npm route.

@JamesGlover , This seems to be a really good suggestion. This way of modifying the tailwind config file allows to choose any of that simple file sharing options including npm. Thank you very much.

**Comment by stevieing** [comment]
Closed as done in favour of #771
